tests/e2e/e2e_test.go

```diff
@@ -192,59 +192,137 @@ func (s *IntegrationTestSuite) TestConcentratedLiquidity() {
+			addrBalances, err := node.QueryBalances(address)
+			s.Require().NoError(err)
+			return addrBalances
+		}
+
+		// Assert balances that are not affected by swap:
+		// * same amount of `stake` in balancesBefore and balancesAfter
+		// * amount of `e2e-default-feetoken` dropped by 1000 (default amount for fee per tx)
+		// * depending on `assertUosmo` and `assertUion` parameters, check that those balances have also not been changed
+		assertDefaultBalances = func(balancesBefore, balancesAfter sdk.Coins, assertUosmo, assertUion bool) {
+			s.Require().True(balancesAfter.AmountOf("stake").Equal(balancesBefore.AmountOf("stake")))
+			s.Require().True(balancesAfter.AmountOf("e2e-default-feetoken").Equal(balancesBefore.AmountOf("e2e-default-feetoken").Sub(defaultFeePerTx)))
+			if assertUion {
+				s.Require().True(balancesAfter.AmountOf("uion").Equal(balancesBefore.AmountOf("uion")))
+			}
+			if assertUosmo {
+				s.Require().True(balancesAfter.AmountOf("uosmo").Equal(balancesBefore.AmountOf("uosmo")))
+			}
+		}
+
+		// calculateFeeGrowthGlobal calculates fee growth global per unit of virtual liquidity based on swap parameters:
+		// amountIn - amount being swapped
+		// swapFee - pool's swap fee
+		// rangeLiquidity - current pool liquidity
+		calculateFeeGrowthGlobal = func(amountIn, swapFee, rangeLiquidity sdk.Dec) sdk.Dec {
+			// First we get total fee charge for the swap (ΔY * swapFee)
+			feeChargeTotal := amountIn.Mul(swapFee)
+
+			// Calculating fee growth global (dividing by pool liquidity to find fee growth per unit of virtual liquidity)
+			feeGrowthGlobal := feeChargeTotal.Quo(rangeLiquidity)
+			return feeGrowthGlobal
+		}
+
+		// calculateFeeGrowthInside calculates fee growth inside range per unit of virtual liquidity
+		// feeGrowthGlobal - global fee growth per unit of virtual liquidity
+		// feeGrowthBelow - fee growth below lower tick
+		// feeGrowthAbove - fee growth above upper tick
+		// Formula: feeGrowthGlobal - feeGrowthBelowLowerTick - feeGrowthAboveUpperTick
+		calculateFeeGrowthInside = func(feeGrowthGlobal, feeGrowthBelow, feeGrowthAbove sdk.Dec) sdk.Dec {
+			return feeGrowthGlobal.Sub(feeGrowthBelow).Sub(feeGrowthAbove)
+		}
+
+		// Helper function for calculating uncollected fees since the time that feeGrowthInsideLast corresponds to
+		// positionLiquidity - current position liquidity
+		// feeGrowthBelow - fee growth below lower tick
+		// feeGrowthAbove - fee growth above upper tick
+		// feeGrowthInsideLast - amount of fee growth inside range at the time from which we want to calculate the amount of uncollected fees
+		// feeGrowthGlobal - variable for tracking global fee growth
+		calculateUncollectedFees = func(positionLiquidity, feeGrowthBelow, feeGrowthAbove, feeGrowthInsideLast sdk.Dec, feeGrowthGlobal sdk.Dec) sdk.Dec {
+			// Calculating fee growth inside range [-1200; 400]
+			feeGrowthInside := calculateFeeGrowthInside(feeGrowthGlobal, feeGrowthBelow, feeGrowthAbove)
```

Comment:
nit: it feels like all these helper funcs (to include the ones above this hi-light) might be better off outside of the test itself. I think tests should be as sparse as possible to not detract from what it is testing, but just an opinion.

Reply:
hmm, makes sense, do you have any suggestions where I could move these to? i do not currently have any ideas, but this suggestion makes sense to me